### PR TITLE
windows support for ansible provisioner

### PR DESCRIPTION
**Pull request by rahulpatil18**
### Summary

This PR is for adding ansible support for windows OS.

The windows inventory file gets created and passes on to ansible in order to connect to windows instance using WINRM connection.

User can pass all WINRM option available in Provisioner connection settings (https://www.terraform.io/docs/provisioners/connection.html)

@frezbo @gkhakare @junaid18183 Please review and suggest the changes.

### CODE REVIEW

File: mode/local_connection_extractor.go [hunk 1]
```diff
@@ -84,9 +93,18 @@ func parseConnectionInfo(s *terraform.InstanceState) (*connectionInfo, error) {
 	if connInfo.Port == 0 {
 		connInfo.Port = DefaultPort
 	}
+
+	if connInfo.Type == "winrm" {
+		if connInfo.Https == true {
+			connInfo.Port = 5986
```
Comment (gkhakare) [inline mode/local_connection_extractor.go]: This should come from the variable. Winrm has two different ports one for http and other https

File: mode/mode_local.go [hunk 2]
```diff
@@ -32,6 +37,59 @@ type inventoryTemplateLocalData struct {
...
+{{" "}}ansible_connection={{.ConnectionType -}}
+{{end -}}
+
+{{if .NTLM }}
+{{" "}}ansible_winrm_transport={{.NTLM -}}
+{{printf "\n" -}}
+{{end -}}
+
+{{if ne .Cacert "" -}}
+{{" "}}ansible_winrm_ca_trust_path={{.Cacert -}}
```
Comment (gkhakare) [inline mode/mode_local.go]: The user will be providing cacert as content in the connection. You should write the cacert content on temporary location and refer that temporary path here

File: mode/local_connection_extractor.go [hunk 3]
```diff
@@ -28,9 +28,20 @@ const (
...
 type connectionInfo struct {
-	User       string
-	Password   string
+	User     string
+	Password string
+	Type     string
+	Https    bool
+	Insecure bool
+	Ntlm     bool
+	Cacert   string
```
Comment (junaid18183) [inline mode/local_connection_extractor.go]: Can we add the json key values to this struct keys ?

File: mode/mode_local.go [hunk 4]
```diff
@@ -57,12 +115,6 @@ const inventoryTemplateLocal = `{{$top := . -}}
 func NewLocalMode(o terraform.UIOutput, s *terraform.InstanceState) (*LocalMode, error) {
 
 	connType := s.Ephemeral.ConnInfo["type"]
-	switch connType {
-	case "ssh", "": // The default connection type is ssh, so if connType is empty use ssh
-	default:
-		return nil, fmt.Errorf("Currently, only SSH connection is supported")
```
Comment (junaid18183) [inline mode/mode_local.go]: We should not remove the error handling, please check if the conninfo is valid

File: mode/mode_local.go [hunk 5]
```diff
@@ -317,68 +378,88 @@ func (v *LocalMode) writePem(pk string) (string, error) {
...
-
 		file, err := ioutil.TempFile(os.TempDir(), "temporary-ansible-inventory")
 		defer file.Close()
 		if err != nil {
 			return "", err
 		}
-
 		v.o.Output(fmt.Sprintf("Writing temporary ansible inventory to '%s'...", file.Name()))
-		if err := ioutil.WriteFile(file.Name(), buf.Bytes(), 0644); err != nil {
+		if err := ioutil.WriteFile(file.Name(), []byte(buf.Bytes()), 0644); err != nil {
```
Comment (junaid18183) [inline mode/mode_local.go]: Why we need this explicit typecasting

**Pull request merged**